src/rules/deadTypeBranchRule.ts

```diff
@@ -0,0 +1,232 @@
+import * as Lint from 'tslint';
+import * as tsutils from 'tsutils';
+import * as ts from 'typescript';
+
+type PerProgramConfig = {
+  tc: ts.TypeChecker,
+  ctx: Lint.WalkContext<void>,
+  // Format: filename|startPosition|endPosition
+  // We use this Map to flag only the highest-level block / expressions.
+  flaggedRanges: Set<string>,
+};
+
+/**
+ * This linter rule tries to find dead branches in programs by looking for
+ * unsatisfiable typing constraints.
+ */
+export class Rule extends Lint.Rules.TypedRule {
+  public static metadata: Lint.IRuleMetadata = {
+    ruleName: 'dead-type-branch',
+    description: 'This linter rule tries to find dead branches in programs ' +
+        'by looking for unsatisfiable typing constraints.',
+    rationale: 'Unsatisfiable type constraits indicate unreachable code.',
+    options: null,
+    optionsDescription: '',
+    type: 'typescript',
+    typescriptOnly: true,
+  };
+
+  applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program):
+      Lint.RuleFailure[] {
+    return this.applyWithFunction(sourceFile, (ctx) => this.walk(ctx, program));
+  }
+
+  /**
+   * Looks for if statements to analyze for dead branches.
+   */
+  private static searchForIfStatements(
```

Comment:
We generally try to avoid static methods on the rule. Consider using function declarations.